### DISCUSSION TITLE
Report BarcodeFormat exactly as specified by SymbologyIdentifier?

### DISCUSSION THREAD

**Discussion by axxel** [zxing-cpp/zxing-cpp#926] [Polls]
Related to #345 I stumbled upon the fact that the reported `Result::format` and the 'code' character of the symbology identifier as specified by the different ISO specs are not in complete agreement. `DataBar` and `DataBarExpanded` as well as `QRCode` and `MicroQRCode` are governed by the same specification and are assigned the same symbology identifier but different `BarcodeFormat`s. I'm questioning whether that makes sens.

**Con:**
 * inconsistency 1: those two bits of info (format vs symbology identifier) are conceptually the same
 * inconsistency 2: we don't distinguish between standard and rectangular DataMatrixes or between compact and full Aztec code either
 * forces to distinguish between the two concepts in the implementation

**Pro:**
 1. even more fine grained filtering through `DecodeHints::format` (minute potential performance benefit if only interested in exactly one variant)
 2. allows to distinguish between the two forms from the client code

Question: Who has a use case for either of the 2 'pro' points? (@vkrause @gitlost @markusfisch @huycn @corbers)

**Comment by oehhar** [comment]
I may give perhaps some background information for DMRE (Data Matrix rectangular extension, www.dmre.info), what is not rectangular Data Matrix.

   *   ISO/IEC16022 Data Matrix has 6 rectangular formats. The Symbology ID is ther same, as for the square versions.
   * ISO/IEC 21471 DMRE has additional 18 rectangular formats. The symbology ID is different (]d7+).

![dm_symbology_ID](https://user-images.githubusercontent.com/12413997/175256180-0f0dbc88-8313-4701-8d4a-acd7ccd115b9.png)

It touk years to convince the present industry to allow to extend Data Matrix.
The industry required, that it is far away from Data Matrix, as they feared about the compatibility of installed scanners. I was happy, to get it in the d Symbology ID category, but new modifier values were unavoidable.
Denso came up with the similar rectangular micro-QR. There, the opposition was less, DMRE had already passed. Due to that, rectangular extended micro-QR and miro-QR share the same symbology ID modifier.

IMHO it is quite important to support this as required by the standard. Sorry for this.

Thank you,
Harald

By the way, great zxing-cpp is missing on www.dmre.info. If someone comes with a suitable text, I may add it.

**Comment by axxel** [comment]
> IMHO it is quite important to support this as required by the standard. Sorry for this.

Thanks for your input. I checked our DM implementation and it does indeed return the correct `]d7+` string for DMRE symbols. My point about inconsistencies still stands, though: we don't distinguish between those two DM variants on the `BarcodeFormat` level, while we do for QRCodes. Mentioning the new rMQR extension is adding to the point: would we then need to add `BarcodeFormat::rMQR`? Feels arbitrary to me.

Regarding www.dmre.info, maybe (following the existing style):

**zxing-cpp: open source c++ barcode scanning library with wrappers for other languages**
The DMRE extension was added by [this pull request](https://github.com/nu-book/zxing-cpp/pull/201) and is included since 1.2.0.

**Comment by oehhar** [comment]
Thank you. [www.dmre.info](www.dmre.info) is updated.

rMQR standard is just published. So, this train may start to run...

**Comment by vkrause** [comment]
> Question: Who has a use case for either of the 2 'pro' points?

we are not using any of this so far

**Comment by axxel** [comment]
I realized that the 4 formats `UPCA`, `UPCE`, `EAN13` and `EAN8` are also governed by one ISO spec and are all reported back with symbology identifier `E`. They are partly distinguished by the modifier character.

I constructed a worst case scenario for the `QRCode` + `MicroQRCode` case and there is indeed a non-trivial performance gain (like 30%) to be had if one would only be interested in `MQR` instead of both. Even though this would be a very untypical usage scenario that quite possibly no-one is taking advantage of yet, I'm hesitant to give up that potential speed gain just like that.

I'm thinking about the concept of 'sub-variants' of formats like `QRCode = MircoQRCode | StandardQRCode | rMQR`, which could indeed be implemented just like the constants `OneDCodes` and `TwoDCodes` with the current `BarcodeFormt` constants infrastructure. Whatever sub-variant of QRCode would be found, the reported format would then be `QRCode`. I can imagine to add a text field to the `Result` with some symbology specific meta-information, like the QRCode version string which would then again allow to distinguish between the different subsets.

My goal here is still to find an API that is simple, consistent and won't be changed anytime soon again. What I have discussed above still feels like not quite fulfilling these requirements. :-/
